File: app/src/main/java/com/worldventures/dreamtrips/wallet/ui/settings/help/feedback/payment/impl/PaymentFeedbackPresenterImpl.java
package com.worldventures.dreamtrips.wallet.ui.settings.help.feedback.payment.impl;


import com.worldventures.core.model.EntityStateHolder;
import com.worldventures.core.modules.picker.service.MediaPickerInteractor;
import com.worldventures.core.modules.infopages.service.FeedbackInteractor;
import com.worldventures.dreamtrips.wallet.service.command.settings.WalletSettingsInteractor;
import com.worldventures.dreamtrips.wallet.service.command.settings.help.PaymentFeedbackCommand;
import com.worldventures.dreamtrips.wallet.service.command.settings.help.SendWalletFeedbackCommand;
import com.worldventures.dreamtrips.wallet.ui.common.base.WalletDeviceConnectionDelegate;
import com.worldventures.dreamtrips.wallet.ui.common.navigation.Navigator;
import com.worldventures.dreamtrips.wallet.ui.settings.help.feedback.base.impl.BaseFeedbackPresenterImpl;
import com.worldventures.dreamtrips.wallet.ui.settings.help.feedback.payment.PaymentFeedbackDelegate;
import com.worldventures.dreamtrips.wallet.ui.settings.help.feedback.payment.PaymentFeedbackPresenter;
import com.worldventures.dreamtrips.wallet.ui.settings.help.feedback.payment.PaymentFeedbackScreen;

import rx.Observable;

public class PaymentFeedbackPresenterImpl extends BaseFeedbackPresenterImpl<PaymentFeedbackScreen> implements PaymentFeedbackPresenter {

   private final PaymentFeedbackDelegate paymentFeedbackDelegate;

   public PaymentFeedbackPresenterImpl(Navigator navigator, WalletDeviceConnectionDelegate deviceConnectionDelegate,
         FeedbackInteractor feedbackInteractor,
         WalletSettingsInteractor walletSettingsInteractor, MediaPickerInteractor mediaPickerInteractor) {
      super(navigator, deviceConnectionDelegate, feedbackInteractor, walletSettingsInteractor, mediaPickerInteractor);
      this.paymentFeedbackDelegate = new PaymentFeedbackDelegate();
   }

   @Override
   public void attachView(PaymentFeedbackScreen view) {
      super.attachView(view);
      observeFormValidation();
      observeUpdateStateAttachments();

      getView().changeActionSendMenuItemEnabled(false);
   }

   private void observeUpdateStateAttachments() {
      getAttachmentsManager().getAttachmentsObservable()
            .compose(getView().bindUntilDetach())
            .subscribe(holder -> {
               final int attachmentsCount = getAttachmentsManager().getAttachments().size();
               getView().changeAddPhotosButtonEnabled(attachmentsCount < MAX_PHOTOS_ATTACHMENT);
            });
   }

   @Override
   public void goBack() {
      getNavigator().goBack();
   }

   private void observeFormValidation() {
      Observable.combineLatest(
            getView().observeMerchantName()
                  .map(this::validForm),
            getAttachmentsManager()
                  .getAttachmentsObservable()
                  .map(holder -> holder.state() != EntityStateHolder.State.PROGRESS)
                  .startWith(true),
            (isMerchantNameValid, isAttachmentsUploadFinished)
                  -> isMerchantNameValid && isAttachmentsUploadFinished)
            .compose(getView().bindUntilDetach())
            .subscribe(enable -> getView().changeActionSendMenuItemEnabled(enable));
   }

   private boolean validForm(CharSequence merchantName) {
      final boolean messageIsEmpty = merchantName.toString().trim().isEmpty();
      return !messageIsEmpty && getAttachmentsManager().getFailedOrPendingAttachmentsCount() <= 0;
   }

   @Override
   public void sendFeedback() {
      getView().changeActionSendMenuItemEnabled(false);

      sendFeedbackCommand(new PaymentFeedbackCommand(
            paymentFeedbackDelegate.createFeedback(getView().getPaymentFeedbackViewModel()),
            paymentFeedbackDelegate.getImagesAttachments(getAttachmentsManager())));
   }

   @Override
   public void handleBackOnDataChangedAction() {
      getView().showBackConfirmDialog();
   }

   @Override
   public void discardChanges() {
      clearAttachments();
      getView().discardViewModelChanges();
      goBack();
   }

   @Override
   protected void handleSuccessSentFeedback() {
      discardChanges();
   }

   @Override
   protected void handleFailSentFeedback(SendWalletFeedbackCommand command, Throwable throwable) {
      getView().changeActionSendMenuItemEnabled(true);
   }
}

File: wallet/src/main/java/com/worldventures/wallet/ui/settings/help/feedback/payment/impl/PaymentFeedbackPresenterImpl.kt
package com.worldventures.wallet.ui.settings.help.feedback.payment.impl


import com.worldventures.core.model.EntityStateHolder
import com.worldventures.wallet.service.command.settings.WalletSettingsInteractor
import com.worldventures.wallet.service.command.settings.help.PaymentFeedbackCommand
import com.worldventures.wallet.service.command.settings.help.SendWalletFeedbackCommand
import com.worldventures.wallet.ui.common.base.WalletDeviceConnectionDelegate
import com.worldventures.wallet.ui.common.navigation.Navigator
import com.worldventures.wallet.ui.settings.help.feedback.base.impl.BaseFeedbackPresenterImpl
import com.worldventures.wallet.ui.settings.help.feedback.base.impl.FeedbackAttachmentsPresenterDelegate
import com.worldventures.wallet.ui.settings.help.feedback.payment.PaymentFeedbackPresenter
import com.worldventures.wallet.ui.settings.help.feedback.payment.PaymentFeedbackScreen

import rx.Observable

class PaymentFeedbackPresenterImpl(navigator: Navigator, deviceConnectionDelegate: WalletDeviceConnectionDelegate,
                                   feedbackAttachmentsPresenterDelegate: FeedbackAttachmentsPresenterDelegate,
                                   walletSettingsInteractor: WalletSettingsInteractor) : BaseFeedbackPresenterImpl<PaymentFeedbackScreen>(navigator, deviceConnectionDelegate, feedbackAttachmentsPresenterDelegate, walletSettingsInteractor), PaymentFeedbackPresenter {

   private val paymentFeedbackConverter: PaymentFeedbackConverter = PaymentFeedbackConverter()

   override fun attachView(view: PaymentFeedbackScreen) {
      super.attachView(view)
      observeFormValidation()
      observeUpdateStateAttachments()
   }

   private fun observeUpdateStateAttachments() {
      attachmentDelegate.attachmentsObservable
            .compose(view!!.bindUntilDetach())
            .subscribe { view!!.changeAddPhotosButtonEnabled(attachmentDelegate.hasFailedOrPendingAttachments) }
   }

   override fun goBack() {
      navigator.goBack()
   }

   override fun showBackConfirmation() {
      view!!.showBackConfirmDialog()
   }

   private fun observeFormValidation() {
      Observable.combineLatest(
            view!!.observeMerchantName()
                  .map<Boolean> { it.isNotBlank() }
                  .startWith(false),
            attachmentDelegate.attachmentsObservable
                  .map { holder -> holder.state() != EntityStateHolder.State.PROGRESS }
                  .startWith(true)
      ) { isMerchantNameValid, isAttachmentsUploadFinished ->
         isMerchantNameValid && isAttachmentsUploadFinished && !attachmentDelegate.hasFailedOrPendingAttachments
      }
            .compose(view!!.bindUntilDetach())
            .subscribe { enable -> view!!.changeActionSendMenuItemEnabled(enable) }
   }

   override fun sendFeedback() {
      view!!.changeActionSendMenuItemEnabled(false)

      sendFeedbackCommand(PaymentFeedbackCommand(
            paymentFeedbackConverter.createFeedback(view!!.paymentFeedbackViewModel),
            attachmentDelegate.imagesAttachments))
   }

   override fun discardChanges() {
      attachmentDelegate.clearAttachments()
      view!!.discardViewModelChanges()
      goBack()
   }

   override fun handleSuccessSentFeedback() {
      discardChanges()
   }

   override fun handleFailSentFeedback(command: SendWalletFeedbackCommand<*>, throwable: Throwable) {
      view!!.changeActionSendMenuItemEnabled(true)
   }
}
